src/main/java/cat/udl/eps/softarch/handler/PurchaseEventHandler.java

```diff
@@ -43,17 +49,18 @@
     @PreAuthorize("hasRole('USER')")
     public void handlePurchasePreCreate(Purchase purchase) {
         String loggedInAs = SecurityContextHolder.getContext().getAuthentication().getName();
+        User purchaser = userRepository.findOne(loggedInAs);
 
         ZonedDateTime now = ZonedDateTime.now();
         purchase.setCreatedAt(now);
 
         // The current purchaser; buying through accepting an offer will have a different handler.
-        purchase.setPurchaser(loggedInAs);
+        purchase.setPurchaser(purchaser);
 
         // A product cannot be purchased by their owner.
         for (Advertisement advertisement: purchase.getAdvertisements()) {
             Advertisement dbAdvertisement = advertisementRepository.findOne(advertisement.getId());
-            Assert.isTrue(!loggedInAs.equals(dbAdvertisement.getOwner()), "You cannot buy your own product.");
+            Assert.isTrue(!purchaser.getName().equals(dbAdvertisement.getOwner()), "You cannot buy your own product.");
```

Comment:
You are comparing the name with an `User` (remember that you have just changed `.getOwner()` to return an `User`).
Either do:
`!purchaser.equals(dbAdvertisement.getOwner())`
or
`!purchaser.getName().equals(dbAdvertisement.getOwner().getName())`